### PR TITLE
First training task

### PR DESCRIPTION
**Pull request by Zaplatynskyi**

### CODE REVIEW

File: exercise_0/main.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+#include <string>
+#include <vector>
+
+struct Result {
+    int vowels;
+    int consonant;
+    int others;
+};
+
+Result countLetter(std::string firstLetter) {
```
Comment (reczepio) [inline exercise_0/main.cpp]: Unnecessary copy. Please use: const std::string& firstLetter

File: exercise_0/main.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    Result result;
+    std::string vowels = "AEIOUYaeouy";
```
Comment (reczepio) [inline exercise_0/main.cpp]: const std::string

File: exercise_0/main.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    int countVowels = 0;
+    int countConsonant = 0;
+    int countOther = 0;
+    for (int i = 0; i < firstLetter.length(); i++) {
```
Comment (reczepio) [inline exercise_0/main.cpp]: first letter can be longer then 1?

File: exercise_0/main.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        if (isalpha(firstLetter[i]))
+            if (vowels.find(firstLetter[i]) != std::string::npos)
+                countVowels += 1;
```
Comment (reczepio) [inline exercise_0/main.cpp]: Instead of this, you can use variable result directly.

File: exercise_0/main.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        else
+            countOther += 1;
+    }
+    result.vowels = countVowels;
+    result.consonant = countConsonant;
+    result.others = countOther;
+    return result;
+}
+
+std::string firstLetterWord(std::string inputText) {
```
Comment (reczepio) [inline exercise_0/main.cpp]: const std::string& inputText

Comment (reczepio) [inline exercise_0/main.cpp]: What idea is behind this function?

File: exercise_0/main.cpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        else if (inputText[i] != ' ' && v == true) {
+            result.push_back(inputText[i]);
+            v = false;
+        }
+    }
+    return result;
+}
+
+
+int main() {
```
Comment (reczepio) [inline exercise_0/main.cpp]: Add argc and argv. Use it for selecting input.

File: exercise_0/main.cpp [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    std::cout << "Enter 1 for read from file or 2  for read from console: " << std::endl;
+    std::cin >> args;
+    if (args == 1) {
+        std::ifstream file("file.txt");
+        if (file.is_open()) {
+            while (getline(file, line)) {
+                firstLetters.append(firstLetterWord(line));
+            }
+            res = countLetter(firstLetters);
+            file.close();
```
Comment (reczepio) [inline exercise_0/main.cpp]: Not needed. Please see RAII pattern

File: exercise_0/main.cpp [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        }
+    }else if (args == 2) {
+        std::cout << "Enter words (q to quit):" << std::endl;
+        while (std::getline(std::cin, line)) {
```
Comment (reczepio) [inline exercise_0/main.cpp]: it reads whole line. Please read word by word (operators "<<" and ">>")

File: exercise_0/main.cpp [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+struct Result {
+    int vowels;
+    int consonant;
+    int others;
+};
+
+Result countLetter(std::string firstLetter) {
+    Result result;
+    std::string vowels = "AEIOUYaeouy";
+    int countVowels = 0;
```
Comment (pstanisz) [inline exercise_0/main.cpp]: You can use the struct fields directly and skip local variables, i.e. `++result.vowels;`

File: exercise_0/main.cpp (hunk 3 above)
Comment (pstanisz) [inline exercise_0/main.cpp]: `++i`

File: exercise_0/main.cpp [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    result.vowels = countVowels;
+    result.consonant = countConsonant;
+    result.others = countOther;
+    return result;
+}
+
+std::string firstLetterWord(std::string inputText) {
+    std::string result = "";
+    bool v = true;
+    for (int i = 0; i < inputText.length(); i++) {
```
Comment (pstanisz) [inline exercise_0/main.cpp]: `++i`

File: exercise_0/main.cpp [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+        if (inputText[i] == ' ')
+            v = true;
+        else if (inputText[i] != ' ' && v == true) {
+            result.push_back(inputText[i]);
+            v = false;
+        }
+    }
+    return result;
```
Comment (pstanisz) [inline exercise_0/main.cpp:25]: Are you searching for first letters of each word to create a single string with them and count then?

File: exercise_0/main.cpp (hunk 7 above)
Comment (pstanisz) [inline exercise_0/main.cpp]: I'd say that it is not needed, but not an error.

File: exercise_0/main.cpp [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+                res = countLetter(firstLetters);
+                break;
+            } else
+                firstLetters.append(firstLetterWord(line));
+        }
+    }
+    std::cout <<res.vowels <<" words beginning with vowels" << std::endl;
+    std::cout <<res.consonant <<" words beginning with consonants " << std::endl;
+    std::cout <<res.others <<" others" << std::endl;
+    return EXIT_FAILURE;
```
Comment (pstanisz) [inline exercise_0/main.cpp]: Why failure? I suppose that it is a success case?

File: exercise_0/main.cpp (hunk 11 above)
Comment (Zaplatynskyi) [reply exercise_0/main.cpp:25]: yes, it does.

File: exercise_0/main.cpp [hunk 13]
```diff
@@ -4,77 +4,53 @@
...
-    std::string vowels = "AEIOUYaeouy";
-    int countVowels = 0;
-    int countConsonant = 0;
-    int countOther = 0;
-    for (int i = 0; i < firstLetter.length(); i++) {
+const std::string vowels = "AEIOUYaeouy";
+
+Result countLetter(const std::string& firstLetter) {
+    Result result{};
+    for (int i = 0; i < firstLetter.length(); ++i) {
```
Comment (pstanisz) [inline exercise_0/main.cpp]: You can also use iterator here.

**Review by pstanisz — APPROVED** [review]

**Review by pstanisz — APPROVED** [review]

**Pull request merged**